arangod/RocksDBEngine/RocksDBOptimizerRules.cpp

```diff
@@ -144,3 +148,142 @@ void RocksDBOptimizerRules::reduceExtractionToProjectionRule(Optimizer* opt,
+      continue;
+    }
+    
+    auto const variable = elements[0].var;
+    TRI_ASSERT(variable != nullptr);
+    
+    auto setter = plan->getVarSetBy(variable->id);
+    
+    if (setter == nullptr || setter->getType() != EN::CALCULATION) {
+      continue;
+    }
+    
+    auto cn = ExecutionNode::castTo<CalculationNode*>(setter);
+    auto const expression = cn->expression();
+    
+    if (expression == nullptr || expression->node() == nullptr ||
+        expression->node()->type != NODE_TYPE_FCALL) {
+      // not the right type of node
+      continue;
+    }
+    
+    auto funcNode = expression->node();
+    auto func = static_cast<Function const*>(funcNode->getData());
+    
+    // we're looking for "RAND()", which is a function call
+    // with an empty parameters array
+    if (func->name != "RAND" || funcNode->numMembers() != 1 ||
+        funcNode->getMember(0)->numMembers() != 0) {
+      continue;
+    }
+    
+    // now we're sure we got SORT RAND() !
+    
+    // we found what we were looking for!
+    // now check if the dependencies qualify
+    if (!n->hasDependency()) {
+      break;
+    }
+    
+    auto current = n->getFirstDependency();
+    ExecutionNode* collectionNode = nullptr;
+    
+    while (current != nullptr) {
+      switch (current->getType()) {
+        case EN::SORT:
+        case EN::COLLECT:
+        case EN::FILTER:
+        case EN::SUBQUERY:
+        case EN::ENUMERATE_LIST:
+        case EN::TRAVERSAL:
+        case EN::SHORTEST_PATH:
+        case EN::INDEX: {
+          // if we found another SortNode, a CollectNode, FilterNode, a
+          // SubqueryNode, an EnumerateListNode, a TraversalNode or an IndexNode
+          // this means we cannot apply our optimization
+          collectionNode = nullptr;
+          current = nullptr;
+          continue;  // this will exit the while loop
+        }
+          
+        case EN::ENUMERATE_COLLECTION: {
+          if (collectionNode == nullptr) {
+            // note this node
+            collectionNode = current;
+            break;
+          } else {
+            // we already found another collection node before. this means we
+            // should not apply our optimization
+            collectionNode = nullptr;
+            current = nullptr;
+            continue;  // this will exit the while loop
+          }
+          // cannot get here
+          TRI_ASSERT(false);
+        }
+          
+        default: {
+          // ignore all other nodes
+        }
+      }
+      
+      if (!current->hasDependency()) {
+        break;
+      }
+      
+      current = current->getFirstDependency();
+    }
+    
+    if (collectionNode == nullptr) {
+      continue; // skip
+    }
+    
+    current = node->getFirstParent();
+    bool valid = false;
+    while (current != nullptr) {
+      if (current->getType() == EN::LIMIT) {
+        LimitNode* ln = ExecutionNode::castTo<LimitNode*>(current);
+        if (ln->limit() == 1) {
```

Comment:
I think this doesn't take into account that there is also an offset, e.g. `LIMIT 1000, 1`. I think in this case it will not work correctly.